llvm/lib/Target/AMDGPU/GCNPreRAOptimizations.cpp

```diff
@@ -225,6 +254,712 @@ bool GCNPreRAOptimizationsImpl::processReg(Register Reg) {
+  // for this use case
+  Op0H_Op1H.addImm(0); // omod
+  LIS->InsertMachineInstrInMaps(*Op0H_Op1H);
+
+  if (I.getFlag(MachineInstr::MIFlag::NoFPExcept)) {
+    Op0L_Op1L->setFlag(MachineInstr::MIFlag::NoFPExcept);
+    Op0H_Op1H->setFlag(MachineInstr::MIFlag::NoFPExcept);
+  }
+  LIS->RemoveMachineInstrFromMaps(I);
+  I.eraseFromParent();
+  LIS->removeInterval(DstReg);
+  LIS->createAndComputeVirtRegInterval(DstReg);
+  return;
+}
+
+void GCNPreRAOptimizationsImpl::processF32Unpacking(MachineInstr &I) {
+  if (I.getOpcode() == AMDGPU::V_PK_FMA_F32) {
+    processFMAF32Unpacking(I);
+    return;
+  }
+  MachineBasicBlock &MBB = *I.getParent();
+  MachineRegisterInfo &MRI = MBB.getParent()->getRegInfo();
+  MachineFunction &MF = *MBB.getParent();
+
+  Register DstReg = I.getOperand(0).getReg();
+  Register SrcReg1 = I.getOperand(2).getReg();
+  Register SrcReg2 = I.getOperand(4).getReg();
+  MachineOperand &DstMO = I.getOperand(0);
+  MachineOperand &SrcMO1 = I.getOperand(2);
+  MachineOperand &SrcMO2 = I.getOperand(4);
+
+  const DebugLoc &DL = I.getDebugLoc();
+  const TargetRegisterClass *DstRC = MRI.getRegClass(I.getOperand(0).getReg());
+  const TargetRegisterClass *Src0RC = MRI.getRegClass(I.getOperand(2).getReg());
+  const TargetRegisterClass *Src1RC = MRI.getRegClass(I.getOperand(4).getReg());
+
+  bool IsVReg64 = (DstRC->getID() == AMDGPU::VReg_64_Align2RegClassID);
+  insertUnpackedF32MI(I, DstMO, SrcMO1, SrcMO2, SrcMO1, SrcMO2, IsVReg64);
+  return;
+}
+
+void GCNPreRAOptimizationsImpl::processF16Unpacking(MachineInstr &I,
+                                                    uint16_t AvailableBudget) {
+  MachineBasicBlock &MBB = *I.getParent();
+  MachineRegisterInfo &MRI = MBB.getParent()->getRegInfo();
+  MachineOperand &DstMO = I.getOperand(0);
+  MachineOperand &SrcMO0 = I.getOperand(2);
+  MachineOperand &SrcMO1 = I.getOperand(4);
+  Register DstReg = DstMO.getReg();
+  Register SrcReg0 = SrcMO0.getReg();
+  Register SrcReg1 = SrcMO1.getReg();
+  const DebugLoc &DL = I.getDebugLoc();
+
+  const TargetRegisterClass *RC = &AMDGPU::VGPR_32RegClass;
+  auto SchedModel = TII->getSchedModel();
+
+  auto BuildImm = [&](uint32_t Val) -> std::pair<Register, uint16_t> {
+    Register ImmReg = MRI.createVirtualRegister(RC);
+    auto NewMI = BuildMI(MBB, I, DL, TII->get(AMDGPU::V_MOV_B32_e32), ImmReg)
+                     .addImm(Val);
+    LIS->InsertMachineInstrInMaps(*NewMI);
+    const MCSchedClassDesc *SchedClassDesc =
+        SchedModel.resolveSchedClass(NewMI);
+    uint16_t LatencyCycles =
+        SchedModel.getWriteProcResBegin(SchedClassDesc)->ReleaseAtCycle;
+    return {ImmReg, LatencyCycles};
+  };
+
+  if (!IsF16MaskSet) {
+    std::pair<Register, uint16_t> RegAndLatency = BuildImm(0x0000FFFF);
+    MaskLo = RegAndLatency.first; // mask for lower 16 bits
+    RegAndLatency = BuildImm(16);
+    ShiftAmt = RegAndLatency.first; // mask for higher 16 bits
+    IsF16MaskSet = true;
```

Comment:
So this just sets MaskLo to the masked Lo bits of the first unpacked F16 instruction? What if the next F16 unpack candidate uses a different register?

Can you add a test where we unpack multiple F16 instructions and they have different source registers? You may need multiple MFMAs.